### PR TITLE
Nuvoton NuMicro M48x: add GPIO and pinctrl drivers, various fixes

### PR DESCRIPTION
**Pull request by fbrozovic**
This PR adds GPIO and pinctrl drivers for the Nuvoton NuMicro MCU family and converts the NuMicro UART driver to use the pinctrl API for configuring its pins. It also adds definitions for the on-board buttons and LEDs on the Nuvoton PFM M487 board.

Depends on zephyrproject-rtos/hal_nuvoton#4 to be merged for the pinctrl dtsi file.

### CODE REVIEW

File: include/zephyr/dt-bindings/gpio/numicro-gpio.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * The debounce flag is a Zephyr specific extension of the standard GPIO flags
+ * specified by the Linux GPIO binding. Only applicable for Nuvoton NuMicro SoCs.
+ */
+#define NUMICRO_GPIO_INPUT_DEBOUNCE	(1U << 8)
+
+#define NUMICRO_GPIO_SLEW_RATE_NORMAL	(0U << NUMICRO_GPIO_SLEW_RATE_SHIFT)
+#define NUMICRO_GPIO_SLEW_RATE_HIGH	(1U << NUMICRO_GPIO_SLEW_RATE_SHIFT)
+#define NUMICRO_GPIO_SLEW_RATE_FAST	(2U << NUMICRO_GPIO_SLEW_RATE_SHIFT)
+#define NUMICRO_GPIO_SLEW_RATE_MASK	(3U << NUMICRO_GPIO_SLEW_RATE_SHIFT)
+#define NUMICRO_GPIO_SLEW_RATE_SHIFT	9
```
Comment (gmarull) [inline include/zephyr/dt-bindings/gpio/numicro-gpio.h:21]: nit: define before

File: drivers/pinctrl/pinctrl_numicro.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ * Copyright (c) 2022 SEAL AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/numicro-pinctrl.h>
+
+#define DT_DRV_COMPAT nuvoton_numicro_pinctrl
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_numicro.c:11]: nit: move to the top by convention

File: drivers/pinctrl/pinctrl_numicro.c [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+	if (pin->input_disable == 0) {
+		flags |= GPIO_INPUT;
+	}
+
+	if (pin->input_debounce != 0) {
+		flags |= NUMICRO_GPIO_INPUT_DEBOUNCE;
+	}
+
+	return gpio_pin_configure(port, NUMICRO_PIN(pin->pinmux), flags);
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_numicro.c:79]: -1 on using GPIO API in pinctrl drivers, please, make them independent even if that involves certain code duplication. This has been the rule for all other drivers. Of course you can still use GPIO DT descriptions if needed.

File: include/zephyr/dt-bindings/gpio/numicro-gpio.h (hunk 1 above)
Comment (gmarull) [inline include/zephyr/dt-bindings/gpio/numicro-gpio.h:21]: looks like this is really meaningful in pinctrl context, will GPIO API (slow) ever care about slew rate?

File: drivers/serial/uart_numicro.c [hunk 4]
```diff
@@ -8,13 +8,19 @@
 #include <zephyr/drivers/uart.h>
 #include <NuMicro.h>
 #include <string.h>
+#ifdef CONFIG_PINCTRL
+#include <zephyr/drivers/pinctrl.h>
+#endif
```
Comment (gmarull) [inline drivers/serial/uart_numicro.c:13]: is pinctrl really optional in these drivers? what other method you have to configure pins? (remember we are about to deprecate pinmux/other legacy methods in 3.2)

File: dts/bindings/serial/nuvoton,numicro-uart.yaml [hunk 5]
```diff
@@ -2,8 +2,11 @@ description: NUVOTON NUMICRO FAMILY UART
...
 
-include: uart-controller.yaml
+include: [uart-controller.yaml, pinctrl-device.yaml]
 
 properties:
     reg:
       required: true
+
+    pinctrl-0:
+      required: true
```
Comment (gmarull) [inline dts/bindings/serial/nuvoton,numicro-uart.yaml:12]: see my previous comment: you are forcing pinctrl, so driver doesn't need to make its use conditional.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/pinctrl/pinctrl_numicro.c (hunk 3 above)
Comment (fbrozovic) [reply drivers/pinctrl/pinctrl_numicro.c:79]: OK - pinctrl_stm32.c e.g. uses a separate function `gpio_stm32_configure` which is defined in gpio_stm32.c. Is that approach acceptable?

File: include/zephyr/dt-bindings/gpio/numicro-gpio.h (hunk 1 above)
Comment (fbrozovic) [reply include/zephyr/dt-bindings/gpio/numicro-gpio.h:21]: Probably not, will move it to the pinctrl driver

File: drivers/pinctrl/pinctrl_numicro.c (hunk 3 above)
Comment (gmarull) [reply drivers/pinctrl/pinctrl_numicro.c:79]: No. That is the only case in-tree, mainly due to legacy reasons and it should be fixed at some point.

File: include/zephyr/dt-bindings/gpio/numicro-gpio.h (hunk 1 above)
Comment (fbrozovic) [reply include/zephyr/dt-bindings/gpio/numicro-gpio.h:21]: ack, will move to pinctrl driver (see other comment)

File: drivers/pinctrl/pinctrl_numicro.c (hunk 2 above)
Comment (fbrozovic) [reply drivers/pinctrl/pinctrl_numicro.c]: ack

File: drivers/serial/uart_numicro.c (hunk 4 above)
Comment (fbrozovic) [reply drivers/serial/uart_numicro.c]: ack, I removed the ifdefs

**Comment by fbrozovic** [comment]
@gmarull I've addressed all of the comments.

File: drivers/pinctrl/pinctrl_numicro.c [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+ * Copyright (c) 2022 SEAL AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numicro_pinctrl
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/numicro-pinctrl.h>
+#include <soc.h>
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_numicro.c:11]: why is this needed? please use `<NuMicro.h>` instead.

As a general rule, include necessary headers instead of relying on indirect includes, some I immediately see:

```
<stdint.h>

<zephyr/devicetree.h>
<zephyr/sys/util.h>
```

File: dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      description: |
+        NuMicro pin controller pin configuration node
+      properties:
+        pinmux:
+          required: true
+          type: array
+          description: |
+            Pin mux selections for this group. See the SoC level pinctrl dtsi file
+            for a defined list of these options.
+        slew-rate:
```
Comment (gmarull) [inline dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml:50]: suggestion: if `normal` is the register default for all pins, you may set a default here as well. Disregard this comment if not applicable.

**Comment by fbrozovic** [comment]
@gmarull Thanks for the review, I went through and included the necessary headers directly, I hope I didn't miss any.

File: dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+          default: "normal"
+          type: string
+          enum:
+            - "normal"
+            - "high"
+            - "fast"
+          description: |
+            Pin output slew rate. Sets the HSRENx register.
```
Comment (gmarull) [inline dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml:57]: you need to justify default here, e.g. it is reset default.

**Review by gmarull — DISMISSED** [review]
lgtm, please merge HAL PR and update sha

**Comment by fbrozovic** [comment]
@carlescufi Could you please have a look at the HAL changes when you have time?

**Comment by carlescufi** [comment]
> @carlescufi Could you please have a look at the HAL changes when you have time?

This is merged now.

**Comment by fbrozovic** [comment]
> This is merged now.

Thanks, SHA updated.

**Review by gmarull — DISMISSED** [review]

**Comment by ssekar15** [comment]
Looks fine. LGTM

**Comment by fbrozovic** [comment]
@gmarull ping

File: drivers/gpio/gpio_numicro.c [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+		} else if ((flags & GPIO_PULL_DOWN) != 0) {
+			bias = GPIO_PUSEL_PULL_DOWN;
+		}
+	}
+
+	regs->MODE = (regs->MODE & ~(3 << (pin * 2))) | (mode << (pin * 2));
+	regs->DBEN = (regs->DBEN & ~BIT(pin)) | (debounce_enable << pin);
+	regs->SMTEN = (regs->SMTEN & ~BIT(pin)) | (schmitt_enable << pin);
+	regs->DINOFF = (regs->DINOFF & ~BIT(pin + 16)) | (disable_input_path << (pin + 16));
+	regs->PUSEL = (regs->PUSEL & ~(3 << (pin * 2))) | (bias << (pin * 2));
```
Comment (gmarull) [inline drivers/gpio/gpio_numicro.c:90]: nit: I see some magic numbers here, can they be defined?

File: drivers/gpio/gpio_numicro.c [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+	return 0;
+}
+
+static int gpio_numicro_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_numicro_config *cfg = dev->config;
+
+	*value = cfg->regs->PIN & 0xffff;
```
Comment (gmarull) [inline drivers/gpio/gpio_numicro.c:99]: same here: define mask (others as well)

File: dts/arm/nuvoton/m48x.dtsi [hunk 11]
```diff
@@ -36,6 +37,78 @@
 	};
 
 	soc {
+		gpioa: gpio@40004000 {
+			compatible = "nuvoton,numicro-gpio";
+			reg = <0x40004000 0x40>;
+			status = "disabled";
+			interrupts = <16 2>;
+			gpio-controller;
```
Comment (gmarull) [inline dts/arm/nuvoton/m48x.dtsi:53]: nit: you could also add `num-gpios` prop

File: dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+          type: string
+          enum:
+            - "normal"
+            - "high"
+            - "fast"
+          description: |
+            Pin output slew rate. Sets the HSRENx register. If not set, defaults to the
+            reset value (normal).
+        input-debounce:
+          required: false
```
Comment (gmarull) [inline dts/bindings/pinctrl/nuvoton,numicro-pinctrl.yaml:60]: required: false is redundant

**Review by gmarull — COMMENTED** [review]
lgtm, a few nitpicks

**Comment by fbrozovic** [comment]
@gmarull comments addressed, please recheck

**Review by gmarull — DISMISSED** [review]

**Comment by fbrozovic** [comment]
@gmarull I also included defines for the magic numbers in the pinctrl driver and fixed a build failure after rebasing, hope this is OK now

**Review by gmarull — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]
LGTM on UART.

**Comment by stephanosio** [comment]
@mnkp Could you take a look?
